File: Assets/Bundles/Client/Resources/TS/Controllers/Global/ProximityPrompt/ProximityPromptController.lua
-- Compiled with unity-ts v2.1.0-75
local Reflect = require("Shared/rbxts_include/node_modules/@easy-games/flamework-core/out/init").Reflect
local Controller = require("Shared/rbxts_include/node_modules/@easy-games/flamework-core/out/init").Controller
local ClientSignals = require("Client/TS/ClientSignals").ClientSignals
local Game = require("Shared/TS/Game").Game
local Keyboard = require("Shared/TS/UserInput/init").Keyboard
local Task = require("Shared/TS/Util/Task").Task
-- * Prompt poll rate, how frequently we update `activatableProximityPrompts`.
local PROMPT_POLL_RATE = 0.1
local ProximityPromptController
do
	ProximityPromptController = setmetatable({}, {
		__tostring = function()
			return "ProximityPromptController"
		end,
	})
	ProximityPromptController.__index = ProximityPromptController
	function ProximityPromptController.new(...)
		local self = setmetatable({}, ProximityPromptController)
		return self:constructor(...) or self
	end
	function ProximityPromptController:constructor()
		self.keyboard = Keyboard.new()
		self.proximityPrompts = {}
		self.activatableProximityPrompts = {}
		self.activatableKeycodes = {}
	end
	function ProximityPromptController:OnStart()
		-- Listen for prompt creation.
		ClientSignals.ProximityPromptCreated:Connect(function(event)
			local _proximityPrompts = self.proximityPrompts
			local _prompt = event.prompt
			table.insert(_proximityPrompts, _prompt)
		end)
		-- Listen for keypresses for prompt activation.
		self.keyboard.KeyDown:Connect(function(event)
			self:HandleKeypress(event.Key)
		end)
		-- Start conditionally displaying prompts.
		self:FindActivatablePrompts()
	end
	function ProximityPromptController:HandleKeypress(key)
		local _activatableProximityPrompts = self.activatableProximityPrompts
		local _arg0 = function(prompt)
			return prompt.data.activationKey == key
		end
		-- ▼ ReadonlyArray.find ▼
		local _result
		for _i, _v in _activatableProximityPrompts do
			if _arg0(_v, _i - 1, _activatableProximityPrompts) == true then
				_result = _v
				break
			end
		end
		-- ▲ ReadonlyArray.find ▲
		local eligiblePrompt = _result
		if eligiblePrompt then
			eligiblePrompt:ActivatePrompt()
		end
	end
	function ProximityPromptController:GetDistanceToPrompt(prompt)
		-- If local character does _not_ have a position, fallback to `math.huge`.
		local _localCharacterPosition = Game.LocalPlayer.Character
		if _localCharacterPosition ~= nil then
			_localCharacterPosition = _localCharacterPosition.gameObject.transform.position
		end
		local localCharacterPosition = _localCharacterPosition
		if not localCharacterPosition then
			return math.huge
		end
		-- Otherwise, return distance.
		local _promptPosition = prompt.data.promptPosition
		return (localCharacterPosition - _promptPosition).magnitude
	end
	function ProximityPromptController:FindActivatablePrompts()
		Task:Spawn(function()
			Task:Repeat(PROMPT_POLL_RATE, function()
				local _proximityPrompts = self.proximityPrompts
				local _arg0 = function(prompt)
					local distToPrompt = self:GetDistanceToPrompt(prompt)
					if distToPrompt <= prompt.data.activationRange then
						local alreadyActive = self:GetActivePromptIndexById(prompt.id) > -1
						local _activatableKeycodes = self.activatableKeycodes
						local _activationKey = prompt.data.activationKey
						local keycodeActive = _activatableKeycodes[_activationKey] ~= nil
						--[[
							* If prompt is already active or prompt with same keycode is active,
							* do nothing. Otherwise, display prompt.
						]]
						if not alreadyActive and not keycodeActive then
							local _activatableKeycodes_1 = self.activatableKeycodes
							local _activationKey_1 = prompt.data.activationKey
							_activatableKeycodes_1[_activationKey_1] = true
							local _activatableProximityPrompts = self.activatableProximityPrompts
							local _prompt = prompt
							table.insert(_activatableProximityPrompts, _prompt)
							self:ShowPrompt(prompt)
						end
					else
						local promptIndex = self:GetActivePromptIndexById(prompt.id)
						local wasActive = promptIndex > -1
						-- If prompt was active, but is now out of range, hide prompt.
						if wasActive then
							local _activatableKeycodes = self.activatableKeycodes
							local _activationKey = prompt.data.activationKey
							_activatableKeycodes[_activationKey] = nil
							table.remove(self.activatableProximityPrompts, promptIndex + 1)
							self:HidePrompt(prompt)
						end
					end
				end
				for _k, _v in _proximityPrompts do
					_arg0(_v, _k - 1, _proximityPrompts)
				end
			end)
		end)
	end
	function ProximityPromptController:ShowPrompt(prompt)
		if prompt.promptGameObject then
			prompt.promptGameObject:SetActive(true)
		end
	end
	function ProximityPromptController:HidePrompt(prompt)
		if prompt.promptGameObject then
			prompt.promptGameObject:SetActive(false)
		end
	end
	function ProximityPromptController:GetActivePromptIndexById(promptId)
		local promptIndex = -1
		do
			local i = 0
			local _shouldIncrement = false
			while true do
				if _shouldIncrement then
					i += 1
				else
					_shouldIncrement = true
				end
				if not (i < #self.activatableProximityPrompts) then
					break
				end
				local promptAtIndex = self.activatableProximityPrompts[i + 1]
				if promptAtIndex.id == promptId then
					promptIndex = i
					break
				end
			end
		end
		return promptIndex
	end
end
-- (Flamework) ProximityPromptController metadata
Reflect.defineMetadata(ProximityPromptController, "identifier", "Bundles/Client/Controllers/Global/ProximityPrompt/ProximityPromptController@ProximityPromptController")
Reflect.defineMetadata(ProximityPromptController, "flamework:implements", { "$:flamework@OnStart" })
Reflect.decorate(ProximityPromptController, "$:flamework@Controller", Controller, { {} })
return {
	ProximityPromptController = ProximityPromptController,
}
-- ----------------------------------
-- ----------------------------------

File: Assets/Typescript~/src/Client/Controllers/Global/ProximityPrompt/ProximityPromptController.ts
import { Controller, OnStart } from "@easy-games/flamework-core";
import { ClientSignals } from "Client/ClientSignals";
import { Game } from "Shared/Game";
import { Keyboard } from "Shared/UserInput";
import { Task } from "Shared/Util/Task";
import { ProximityPrompt } from "./ProximityPrompt";

/** Prompt poll rate, how frequently we update `activatableProximityPrompts`. */
const PROMPT_POLL_RATE = 0.1;

@Controller({})
export class ProximityPromptController implements OnStart {
	/** Keyboard instance. */
	private keyboard = new Keyboard();
	/** All active proximity prompts in world. */
	private proximityPrompts: ProximityPrompt[] = [];
	/** Proximity prompts in activation range. */
	private activatableProximityPrompts: ProximityPrompt[] = [];
	/** Key codes in activation range. */
	private activatableKeycodes = new Set<Key>();

	OnStart(): void {
		/* Listen for prompt creation. */
		ClientSignals.ProximityPromptCreated.Connect((event) => {
			this.proximityPrompts.push(event.prompt);
		});
		/* Listen for keypresses for prompt activation. */
		// this.keyboard.KeyDown.Connect((event) => {
		// 	this.HandleKeypress(event.Key);
		// });
		/* Start conditionally displaying prompts. */
		this.FindActivatablePrompts();
	}

	/** Handle keypresses and activate prompts if applicable. */
	// private HandleKeypress(key: Key): void {
	// 	const eligiblePrompt = this.activatableProximityPrompts.find((prompt) => prompt.data.activationKey === key);
	// 	if (eligiblePrompt) {
	// 		eligiblePrompt.ActivatePrompt();
	// 	}
	// }

	/** Returns distance between local player and a proximity prompt. */
	private GetDistanceToPrompt(prompt: ProximityPrompt): number {
		/* If local character does _not_ have a position, fallback to `math.huge`. */
		const localCharacterPosition = Game.LocalPlayer.Character?.gameObject.transform.position;
		if (!localCharacterPosition) return math.huge;
		/* Otherwise, return distance. */
		return localCharacterPosition.sub(prompt.data.promptPosition).magnitude;
	}

	/** Displays and hides prompts based on `activationRange`. */
	private FindActivatablePrompts(): void {
		Task.Spawn(() => {
			Task.Repeat(PROMPT_POLL_RATE, () => {
				this.proximityPrompts.forEach((prompt) => {
					const distToPrompt = this.GetDistanceToPrompt(prompt);
					if (distToPrompt <= prompt.data.activationRange) {
						const alreadyActive = this.GetActivePromptIndexById(prompt.id) > -1;
						// const keycodeActive = this.activatableKeycodes.has(prompt.data.activationKey);
						/*
						 * If prompt is already active or prompt with same keycode is active,
						 * do nothing. Otherwise, display prompt.
						 */
						if (!alreadyActive) {
							// this.activatableKeycodes.add(prompt.data.activationKey);
							this.activatableProximityPrompts.push(prompt);
							this.ShowPrompt(prompt);
							prompt.SetCanActivate(true);
						}
					} else {
						const promptIndex = this.GetActivePromptIndexById(prompt.id);
						const wasActive = promptIndex > -1;
						/* If prompt was active, but is now out of range, hide prompt. */
						if (wasActive) {
							// this.activatableKeycodes.delete(prompt.data.activationKey);
							this.activatableProximityPrompts.remove(promptIndex);
							this.HidePrompt(prompt);
							prompt.SetCanActivate(false);
						}
					}
				});
			});
		});
	}

	/** Shows a proximity prompt. */
	private ShowPrompt(prompt: ProximityPrompt): void {
		if (prompt.promptGameObject) {
			prompt.promptGameObject.SetActive(true);
		}
	}

	/** Hides a proximity prompt. */
	private HidePrompt(prompt: ProximityPrompt): void {
		if (prompt.promptGameObject) {
			prompt.promptGameObject.SetActive(false);
		}
	}

	/**
	 * Returns an active proximity prompt's index.
	 * @param promptId An active proximity prompt id.
	 * @returns Index that corresponds to active prompt with `promptId`. If prompt is _not_ active, the function returns -1.
	 */
	private GetActivePromptIndexById(promptId: string): number {
		let promptIndex = -1;
		for (let i = 0; i < this.activatableProximityPrompts.size(); i++) {
			const promptAtIndex = this.activatableProximityPrompts[i];
			if (promptAtIndex.id === promptId) {
				promptIndex = i;
				break;
			}
		}
		return promptIndex;
	}
}
